protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,713 @@
+        if servers.len() < max_peers && extend_with_static {
+            servers.extend(self.static_servers.clone())
+        }
+        if servers.len() < max_peers && extend_with_connected {
+            let mut connected = self.connected.iter();
+            // TODO: use random set
+            for _ in 0..connected.len() {
+                let peer = connected.next().unwrap();
+                servers.push(*peer);
+                if servers.len() >= max_peers {
+                    break;
+                }
+            }
+        }
+        Probe {
+            server_count: servers.len(),
+            pending_servers: servers,
+            addresses: HashMap::new(),
+            required_success,
+            dismissed: 0,
+            errors: Vec::new(),
+            min_valid,
+        }
+    }
+
+    fn do_probe(&mut self, probe: Probe, addr_records: Vec<AddressRecord>) {
+        let addrs: Vec<Multiaddr> = addr_records.into_iter().map(|r| r.addr).collect();
+        for peer_id in probe.pending_servers.clone() {
+            self.inner.send_request(
+                &peer_id,
+                DialRequest {
+                    peer_id,
+                    addrs: addrs.clone(),
+                },
+            );
+        }
+        let _ = self.ongoing_outbound.insert(probe);
+    }
+
+    // Handle the inbound request and collect the valid addresses to be dialed.
+    fn handle_request(&mut self, sender: PeerId, request: DialRequest) -> Option<Vec<Multiaddr>> {
+        let config = self
+            .server_config
+            .as_ref()
+            .expect("Server config is present.");
+
+        // Validate that the peer to be dialed is the request's sender.
+        if request.peer_id != sender {
+            return None;
+        }
+        // Check that there is no ongoing dial to the remote.
+        if self.ongoing_inbound.contains_key(&sender) {
+            return None;
+        }
+        // Check if max simultaneous autonat dial-requests are reached.
+        if self.ongoing_inbound.len() >= config.max_ongoing {
+            return None;
+        }
+
+        let known_addrs = self.inner.addresses_of_peer(&sender);
+        // At least one observed address was added, either in the `RequestResponse` protocol
+        // if we dialed the remote, or in Self::inject_connection_established if the
+        // remote dialed us.
+        let observed_addr = known_addrs.first().expect("An address is known.");
+        // Filter valid addresses.
+        let mut addrs = filter_valid_addrs(sender, request.addrs, observed_addr);
+        addrs.truncate(config.max_addresses);
+
+        if addrs.is_empty() {
+            return None;
+        }
+
+        Some(addrs)
+    }
+
+    fn handle_response(
+        &mut self,
+        sender: PeerId,
+        response: Option<DialResponse>,
+    ) -> Option<NatStatus> {
+        let mut probe = self.ongoing_outbound.take()?;
+        if !probe.pending_servers.contains(&sender) {
```

Comment:
```suggestion
        // If response is from a peer that is not part of the ongoing probe, ignore it.
        if !probe.pending_servers.contains(&sender) {
```

This took me a bit to understand. How about adding the comment above?